hbase-backup/src/main/java/org/apache/hadoop/hbase/backup/impl/BackupManager.java

```diff
@@ -295,51 +298,31 @@ public ArrayList<BackupImage> getAncestors(BackupInfo backupInfo) throws IOExcep
         .withRootDir(backup.getBackupRootDir()).withTableList(backup.getTableNames())
         .withStartTime(backup.getStartTs()).withCompleteTime(backup.getCompleteTs()).build();
 
-      // Only direct ancestors for a backup are required and not entire history of backup for this
-      // table resulting in verifying all of the previous backups which is unnecessary and backup
-      // paths need not be valid beyond the lifetime of a backup.
-      //
-      // RootDir is way of grouping a single backup including one full and many incremental backups
+      // If the image has a different rootDir, it cannot be an ancestor.
       if (!image.getRootDir().equals(backupInfo.getBackupRootDir())) {
         continue;
       }
 
-      // add the full backup image as an ancestor until the last incremental backup
-      if (backup.getType().equals(BackupType.FULL)) {
-        // check the backup image coverage, if previous image could be covered by the newer ones,
-        // then no need to add
-        if (!BackupManifest.canCoverImage(ancestors, image)) {
-          ancestors.add(image);
-        }
+      // The ancestors consist of the most recent FULL backups that cover the list of tables
+      // required in the new backup and all INCREMENTAL backups that came after one of those FULL
+      // backups.
+      if (backup.getType().equals(BackupType.INCREMENTAL)) {
```

Comment:
I believe we're just iterating through all backup history at this point — what if the given incremental backup is not relevant to the `tablesToCover`? I think that should not be included as an ancestor, and I think that's what the `BackupManifest#canCoverImage` was hoping to achieve (but failing to due to the bugs explained in https://issues.apache.org/jira/browse/HBASE-28562?focusedCommentId=17843008&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17843008

Hyperbolically, on a cluster with hundreds of tables taking incremental backups everyday, this would quickly balloon into huge BackupManifests that cannot fit in a reasonably sized heap

Reply:
I'm not sure I understand your point, or I'm still missing something in my understanding of the backup mechanism.

In my understanding, a newly created incremental backup will include all tables that have been included in a full backup at some point. (I base this on what is shown in `hbase backup history`.) So that means that all incremental backups up to (and including) the original full backups are in fact ancestors.

Reply:
An incremental backup can apply to any subset of the total tables on a cluster, not necessarily all tables. So if your cluster has tables A, B, and C and you're fetching the ancestry for a table C backup, then you really don't want to include ancestors that only apply to table A and/or B

Reply:
I did the following experiment:

```
echo "create 'table1', 'cf'" | bin/hbase shell -n
echo "create 'table2', 'cf'" | bin/hbase shell -n
echo "create 'table3', 'cf'" | bin/hbase shell -n
bin/hbase backup create full file:/tmp/hbasebackups -t table1,table2
echo "put 'table1', 'row1', 'cf:a', 'valueA'" | bin/hbase shell -n
echo "put 'table2', 'row1', 'cf:a', 'valueB'" | bin/hbase shell -n
bin/hbase backup create incremental file:/tmp/hbasebackups -t table1
bin/hbase backup history

{ID=backup_1714763143839,Type=INCREMENTAL,Tables={table2,table1},State=COMPLETE,Start time=Fri May 03 21:05:44 CEST 2024,End time=Fri May 03 21:05:47 CEST 2024,Progress=100%}
{ID=backup_1714763125413,Type=FULL,Tables={table2,table1},State=COMPLETE,Start time=Fri May 03 21:05:25 CEST 2024,End time=Fri May 03 21:05:29 CEST 2024,Progress=100%}
```

This seems to suggest that the incremental backup includes both tables, even though I only requested table1. So I don't think it's possible to create an incremental backup of just a single table if your FULL backups have more tables covered?

Reply:
Interesting... sounds like yet another bug 😄 I'm going to need to read more of the code to understand what is happening. But I believe we _should_ be able to take incremental backups of individual tables — the API seems to allow for it, and it's a clearly useful feature when the value proposition of the entire incremental backup system is to increase the granularity of backup operations

Reply:
I thought this was a consequence of the incremental backups using the WAL to track changes. Incremental backups are created using the WAL-records since the previous full backup. So I figured that any incremental backup would fully consume that historic WAL so disk space is reclaimed. Assuming the WAL would be cleaned up after creating the incremental backup, that means you can't have multiple incremental backups spanning the same time period, hence it would make sense to include all tables in the incremental backup that create.

Some experiments last week (as result of your other feedback), learned me that using different root folders, it is possible to have incremental backups spanning the same time period. So perhaps I was wrong about my other idea as well.

I'll also browse some source code to find a lead on this, but isn't there a technical design written somewhere about how backup-restore is supposed to work?

Reply:
Also relevant: in https://github.com/apache/hbase/pull/5871/ I clean up some usages of `getAncestors`, perhaps it's best to make this method non-public (and keep the current implementation where it only works for currently-being-created backups).

Reply:
For an incremental backup, `BackupAdminImpl#backupTables` throws away the user-specified tables, and replaces them with the output of `BackupSystemTable#getIncrementalBackupTableSet(targetRootDir)`.

There is one "incremental backup set" tracked for backup rootdir. Any table that part of a full backup gets added, entries only get deleted if all backups of that table get deleted. (Note that I just logged https://issues.apache.org/jira/browse/HBASE-28568 for a bug in that logic.)

So it seems the code is intended to work as I originally described.

Reply:
Interesting, so the incremental backups are identical regardless of the table set input from the user? Maybe this API needs a refactor...

Reply:
I'm not so certain that we made the right call here. Experimenting with incremental backups at my day job, and taking an incremental backup for one table versus another (on the same cluster) results in totally different set of file contents for each. I'm still not an expert in this code from beginning to end, but we might have been too quick to assume that the API was lying

Reply:
Hi @rmdmattingly, did you verify differences when restoring those backups? If so, can you share a reproducable set of steps?

Reply:
I'll take a look today to see if I can give reproducible steps. Separately, +1 to the ancestry calculation improvements here — regardless of whether we support table specific ancestry trees, I've been playing around with this changeset vs 2.6 today and observed the several bug fixes that you've implemented here (stopping at the appropriate full backup, no more recursion issues when calculating ancestors' ancestors)

Reply:
I wrote up [this test](https://github.com/HubSpot/hbase/compare/hubspot-2.6...HubSpot:hbase:inc-backup-tablesets) which only proves your point that incremental backups ignore the tableset input — ideally, from my perspective, this test would fail at 
```java
      // contains all 200 rows??? why/how/are they correct
      assertEquals(rowCount, 200);
```
or the API wouldn't accept a set of tables to backup

As for why I'm seeing different backup contents at my day job, it's a good question and probably not as easy to reproduce in a unit test. I'm picking up a years long backup system refactor after a colleague's departure, so I probably won't have a quick answer there either, but will continue to dig.

All this to say, you're correct here and I want to reiterate the improvements I mentioned above, so I think we should ship these changes 🚢

Reply:
@rmdmattingly do we need to add your new unit test someplace?

Reply:
Chatted with Nick this morning. I think there's room for a unit test of this genre — once we figure out exactly what the intention of the TableSet is on the backups API, then we should add a test which enforces these intentions. But this test in its current form isn't useful for that